aws-codeartifact-repository/src/main/java/software/amazon/codeartifact/repository/BaseHandlerStd.java

```diff
@@ -30,4 +44,84 @@
     final ProxyClient<CodeartifactClient> proxyClient,
     final Logger logger);
+
+  protected ProgressEvent<ResourceModel, CallbackContext> associateExternalConnections(
+      final ProgressEvent<ResourceModel, CallbackContext> progress,
+      final CallbackContext callbackContext,
+      final ResourceHandlerRequest<ResourceModel> request,
+      final ProxyClient<CodeartifactClient> proxyClient,
+      final Set<String> externalConnectionsToAdd,
+      final Logger logger
+  ) {
+      ResourceModel resourceModel = request.getDesiredResourceState();
+      logger.log(String.format("Adding external connections: %s", externalConnectionsToAdd.toString()));
+      if (CollectionUtils.isNullOrEmpty(externalConnectionsToAdd)) {
+          logger.log("No external connections to add");
+          // Nothing to add, continue
+          return ProgressEvent.progress(resourceModel, callbackContext);
+      }
+
+      if (externalConnectionsToAdd.size() > 1) {
+          return ProgressEvent.failed(resourceModel, callbackContext, HandlerErrorCode.InvalidRequest,
+              "ExternalConnections for repositories are currently limited to 1.");
+      }
+
+      // Loop is currently not necessary because only 1 external connection is allowed, leaving this in for
+      // when multiple are supported
+      externalConnectionsToAdd.forEach(ec -> {
+          try {
+              AssociateExternalConnectionRequest associateExternalConnectionRequest
+                  = Translator.translateAssociateExternalConnectionsRequest(resourceModel, ec);
+
+              proxyClient.injectCredentialsAndInvokeV2(associateExternalConnectionRequest, proxyClient.client()::associateExternalConnection);
+          } catch (final AwsServiceException e) {
+              String repositoryName = progress.getResourceModel().getRepositoryName();
+              Translator.throwCfnException(e, Constants.ASSOCIATE_EXTERNAL_CONNECTION, repositoryName);
+          }
+          logger.log(String.format("Successfully associated external connection: %s", ec));
+      });
+
+      return ProgressEvent.<ResourceModel, CallbackContext>builder()
+          .resourceModel(resourceModel)
+          .status(OperationStatus.IN_PROGRESS)
+          .build();
+  }
+
+  protected ProgressEvent<ResourceModel, CallbackContext> putRepositoryPermissionsPolicy(
+      final AmazonWebServicesClientProxy proxy,
+      final ProgressEvent<ResourceModel, CallbackContext> progress,
+      final CallbackContext callbackContext,
+      final ProxyClient<CodeartifactClient> proxyClient,
+      final Logger logger
+  ) {
+      final ResourceModel desiredModel = progress.getResourceModel();
+      if (desiredModel.getPolicyDocument() == null) {
+          return ProgressEvent.progress(desiredModel, callbackContext);
+      }
+      return proxy.initiate("AWS-CodeArtifact-Repository::Update::PutRepositoryPermissionsPolicy", proxyClient, progress.getResourceModel(), progress.getCallbackContext())
```

Comment:
IMO, we shouldn't be doing all of these operations under a return statement and break out all of this logic.

Reply:
It does appear a put a verbose, but seems to be in line with the progress chaining framework provided by CloudFormation. Here is the example they provide: 


https://docs.aws.amazon.com/cloudformation-cli/latest/userguide/resource-type-develop-stabilize.html

```
public class CreateHandler extends BaseKinesisHandler {
    //
    // The handler is provide with a AmazonWebServicesClientProxy that provides 
    // the framework for making calls that returns a ProgressEvent, 
    // which can then be chained to perform the next task.
    //
    protected ProgressEvent<ResourceModel, CallbackContext> 
        handleRequest(final AmazonWebServicesClientProxy proxy,
                      final ResourceHandlerRequest<ResourceModel> request,
                      final CallbackContext callbackContext,
                      final ProxyClient<KinesisClient> client,
                      final Logger logger) {

        ResourceModel model = request.getDesiredResourceState();
        if (model.getName() == null || model.getName().isEmpty()) {
            model.setName(
               IdentifierUtils.generateResourceIdentifier(
                  "stream-", request.getClientRequestToken(), 128));
        }
        //
        // 1) initiate the call context, we are making createStream API call 
        // 
        return proxy.initiate(
            "kinesis:CreateStream", client, model, callbackContext)
            
            //
            // 2) transform Resource model properties to CreateStreamRequest API
            //
            .request((m) ->
  ...
```

I could work on cleaning this up though, for example I should be passing in the callbackContext rather than grabbing it using `progress.getCallbackContext()`. I should probably break out line 101 into multiple lines